### DISCUSSION TITLE
Macro versus function versus inline in vim sources

### DISCUSSION THREAD

**Discussion by errael** [vim/vim#13565] [General]
Putting together a PR with a simple macro
```
#define INT_TRIM(x) ((x) > INT_MAX ? INT_MAX : (x) < INT_MIN ? INT_MIN : (x))
```
But it's hard to use because it references `x` three times.
An inline function would be best, but `:help style-compiler` explicitly says
do not use `inline` and "let the optimizer do its work".

If I've got
```
    int
trim_to_int(long long x)
{
    return INT_TRIM(x);
}
```
in a *.c file, how can this be inlined in another *.c file?

Looking at gcc docs, I see a `-flto` option
```
This option runs the standard link-time optimizer.
```
But I don't see this used anywhere in makefiles.

Could do "static int func() {...}" in a header, that might work, but haven't
seen that pattern around (which doesn't mean much given my unfamiliarity with
the vim sources).

I suppose it doesn't really matter, but I dislike adding function call overhead to something so simple.

**Comment by errael** [comment]
**Link time optimization**

In `:help style-compiler` there's
```
Not to be used
...
- "inline" (it's hardly ever needed, let the optimizer do its work)
```
But there is `link time optimization` available with both gcc and clang. `link time optimization` allows automatic inlining of functions from a different "C" file. I don't see this mentioned in `Makefile`. Here's some comparisons exploring link time optimizations. These are based on micro profiling. Results from batch editing/scripting would be interesting.

The two versions of vim used with these tests are created using `gcc -O2` without and with the `-flto` flag to test `link time optimization`.


| speed increase  | vim9script code |
| :---: | --- |
| 18% | the_dict->extend({[a_key]: a_value}) |
| 10% | local = a_class.Getter() |
| 13% | map(l, 'v:val * 2') |
| 6% | map(l, (_, val) => val * 2) |
| 10% | map(l, (_, val: number): number => val * 2) |

**surprising improvement**
For the following, the percentage is the speedup you get by specifying the type.
Note: the % difference is specifying the type versus not specifying the type.
What's curious is that specifying the type helps twice as much with the link optimized vim.
These are the two lines use for this comparison.
```
map(l, (_, val) => val * 2)
map(l, (_, val: number): number => val * 2)
```
It's 4% faster for regular vim; and 8% faster with link optimized vim.

**Notes**
Did `LDFLAGS = $(CFLAGS)` to insure same flags for compile and link.
Using `-flto=auto` makes a huge difference in link time.
Did not use `-fwhole-program` or a linker plugin (AFAICT).
The profiling results were generated using [micro profiling](https://github.com/errael/vim-microprof).

**Reply by ychin** [reply]
FWIW I have already turned on link-time optimization for MacVim binary builds, along with using `-O3` optimization flags. I did some profiling for the PR here: https://github.com/macvim-dev/macvim/pull/1444#issuecomment-1766065232. They were more end-to-end tests than the single-line vim9script micro-benchmarks here. It definitely showed measurable improvements in some cases. I do wonder if it varies across compilers, so I'm not sure if this is something we should turn on for all users or not (since Vim has a large variety of users and some people use ).

If you follow the link, I also linked to the Neovim issue where they also turned this on on their side as well.

As I mentioned on the other comment though, I think the proper way is to just use the `inline` keyword and define the function definition at the header. Whether to use LTO or not is another matter.

**Comment by errael** [comment]
Size stats (forgot to include)
**without -flto**
```
   text	   data	    bss	    dec	    hex	filename
3820176	 283120	  63256	4166552	 3f9398	vim
```
**with -flto**
```
   text	   data	    bss	    dec	    hex	filename
4112691	 282902	  63352	4458945	 4409c1	vim
```
Around 8% bigger `text` using `-flto`

**Comment by ychin** [comment]
>  An inline function would be best, but :help style-compiler explicitly says do not use inline and "let the optimizer do its work".

I feel that this should just be relaxed. These days the reason why `inline` exists is mostly to allow declaring function definitions in a header without triggering a linker error. The compiler will still get to decide whether to inline or not.

**Comment by errael** [comment]
Issue https://github.com/vim/vim/issues/14012 turns out to be about -flto problems. And see https://github.com/vim/vim/issues/14012#issuecomment-1942043795

**Reply by ychin** [reply]
FWIW this seems more an issue of mixing and matching `-flto` flags between compilation and linking steps, due to how the Perl cpp flags were generated. I don't think it affects how LTO works in general.

**Reply by errael** [reply]
I think so as well. And there may also be a difference between clang and gcc in play.
